storage/umass/umass.c

```diff
@@ -0,0 +1,496 @@
+		fprintf(stderr, "umass: devAlloc failed\n");
+		return -ENOMEM;
+	}
+
+	dev->instance = *insertion;
+	if ((dev->pipeCtrl = usb_open(insertion, usb_transfer_control, 0)) < 0) {
+		free(dev);
+		fprintf(stderr, "umass: usb_open failed\n");
+		return -EINVAL;
+	}
+
+	if (usb_setConfiguration(dev->pipeCtrl, 1) != 0) {
+		free(dev);
+		fprintf(stderr, "umass: setConfiguration failed\n");
+		return -EINVAL;
+	}
+
+	if ((dev->pipeIn = usb_open(insertion, usb_transfer_bulk, usb_dir_in)) < 0) {
+		fprintf(stderr, "umass: pipe open failed \n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	if ((dev->pipeOut = usb_open(insertion, usb_transfer_bulk, usb_dir_out)) < 0) {
+		fprintf(stderr, "umass: pipe open failed\n");
+		free(dev);
+		return -EINVAL;
+	}
+	dev->tag = 0;
+
+	if (umass_check(dev)) {
+		fprintf(stderr, "umass: umass_check failed\n");
+		free(dev);
+		return -EINVAL;
+	}
+
+	snprintf(dev->path, sizeof(dev->path), "/dev/umass%d", dev->id);
+	oid.port = umass_common.msgport;
+	oid.id = dev->id + 1;
+	if (create_dev(&oid, dev->path) != 0) {
+		free(dev);
+		fprintf(stderr, "usb: Can't create dev!\n");
+		return -EINVAL;
+	}
+
+	LIST_ADD(&umass_common.devices, dev);
+	fprintf(stderr, "umass: New USB Mass Storage device: %s sectors: %d\n", dev->path, dev->partSize);
+
+	return 0;
+}
+
+
+static int umass_handleDeletion(usb_deletion_t *del)
+{
+	umass_dev_t *next, *dev = umass_common.devices;
+
+	if (dev == NULL)
+		return 0;
+
+	do {
+		next = dev->next;
+		if (dev->instance.bus == del->bus && dev->instance.dev == del->dev &&
+				dev->instance.interface == del->interface) {
+			remove(dev->path);
+			LIST_REMOVE(&umass_common.devices, dev);
+			fprintf(stderr, "umass: Device removed: %s\n", dev->path);
+			free(dev);
+			if (dev == next)
+				break;
+		}
+		dev = next;
+	} while (dev != umass_common.devices);
+
+	return 0;
+}
+
+
+int main(int argc, char *argv[])
+{
+	int ret, i;
+	msg_t msg;
+	usb_msg_t *umsg = (usb_msg_t *)msg.i.raw;
+
+	/* Port for communication with the USB stack */
+	if (portCreate(&umass_common.drvport) != 0) {
+		fprintf(stderr, "umass: Can't create port!\n");
+		return -ENOMEM;
+	}
+
+	/* Port for communication with driver clients */
+	if (portCreate(&umass_common.msgport) != 0) {
+		fprintf(stderr, "umass: Can't create port!\n");
+		return -ENOMEM;
+	}
+
+	if ((usb_connect(filters, sizeof(filters) / sizeof(filters[0]), umass_common.drvport)) < 0) {
+		fprintf(stderr, "umass: Fail to connect to usb host!\n");
+		return -EINVAL;
+	}
+
+	if (mutexCreate(&umass_common.lock)) {
+		fprintf(stderr, "umass: Can't create mutex!\n");
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < UMASS_N_MSG_THREADS; i++) {
+		if ((ret = beginthread(umass_msgthr, 4, umass_common.stack[i], sizeof(umass_common.stack[i]), NULL)) != 0) {
+			fprintf(stderr, "umass: fail to beginthread ret: %d\n", ret);
+			return -1;
+		}
+	}
+
+	for (;;) {
+		ret = usb_eventsWait(umass_common.drvport, &msg);
+		if (ret != 0)
+			continue;
```

Comment:
Can cause tight loop and high CPU usage on persistent error - maybe we should check what error we got and if this error can be or should be corrected in any way?